cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+    if (availableArtifacts.isEmpty()) {
+      // This should not be possible as at least the current used artifact should be there.
+      LOG.error("No artifacts found for artifact {} ", currentArtifact);
+      throw new InternalError("Issues in trying to find application artifact for upgrading app " + appId);
+    }
+    ArtifactSummary candidateArtifact = availableArtifacts.get(availableArtifacts.size() - 1);
+    ArtifactVersion candidateArtifactVersion = new ArtifactVersion(candidateArtifact.getVersion());
+    if (candidateArtifactVersion.getVersion() == null) {
+      throw new InvalidArtifactException(String.format(
+          "Requested artifact version '%s' is invalid", candidateArtifact.getVersion()));
+    }
+
+    // Check conditions for validity of candidate artifact such as name and scope should be same. Also check that
+    // candidate artifact should have higher version than current artifact.
+    // In ideal cases, these conditions should never happen.
+    if (!currentArtifact.getName().equals(candidateArtifact.getName())) {
+      throw new InvalidArtifactException(String.format(
+          "Requested artifact name %s does not match with current name %s.", candidateArtifact.getName(),
+          currentArtifact.getName()));
+    }
+    if (currentArtifact.getScope() != candidateArtifact.getScope()) {
+      throw new InvalidArtifactException(String.format(
+          "Requested artifact scope %s does not match with current name %s.", candidateArtifact.getScope(),
+          currentArtifact.getScope()));
+    }
+    if (currentArtifact.getVersion().compareTo(candidateArtifactVersion) > 0) {
+      throw new InvalidArtifactException(String.format(
+          "Requested artifact version %s is older than current artifact version %s.", candidateArtifact.getVersion(),
+          currentArtifact.getVersion()));
+    }
+
+    ArtifactId newArtifactId =
+        new ArtifactId(currentArtifact.getName(), candidateArtifactVersion, currentArtifact.getScope());
+
+    Id.Artifact newArtifact = Id.Artifact.fromEntityId(Artifacts.toProtoArtifactId(appId.getParent(), newArtifactId));
+    ArtifactDetail newArtifactDetail = artifactRepository.getArtifact(newArtifact);
+    List<ApplicationConfigUpdateAction> upgradeActions = Arrays.asList(ApplicationConfigUpdateAction.UPGRADE_ARTIFACT);
+
+    try {
+      ApplicationUpdateDetails detail =
+          updateApplicationInternal(appId, appId.getParent(), appId.getApplication(), null,
+                                    currentSpec.getConfiguration(), programTerminator, newArtifactDetail,
+                                    upgradeActions, ownerAdmin.getOwner(appId), /*updateSchedules=*/false);
+      LOG.debug("Application upgrade successful. Update details: {}. Error: {}", detail.getUpdateDetails(),
+                detail.getError());
+      return new ApplicationUpdateDetails("upgrade successful.", null);
+    } catch (UnsupportedOperationException E) {
```

Comment:
E -> e

Reply:
also, I thought there was some boolean that can be used to check if update is supported? 

UnsupportedOperationException can be thrown by other things (for example, trying to write to an UnmodifiableMap), so it's better to base this off the API